wandb/sdk/launch/docker.py

```diff
@@ -45,6 +47,207 @@ def validate_docker_installation() -> None:
         )
 
 
+def get_docker_user(launch_project):
+    import getpass
+
+    username = getpass.getuser()
+    userid = launch_project.docker_user_id or os.geteuid()
+    return username, userid
+
+
+TEMPLATE = """
+FROM {py_build_image} as build
+
+# install in venv to copy
+RUN python -m venv /opt/venv
+ENV PATH="/opt/venv/bin:$PATH"
+
+COPY src/requirements.txt .
+# different requirements line if we have buildx or not
+{requirements_line}
+
+# different base image for cpu/gpu
+{base_setup}
+
+COPY --from=build /opt/venv /opt/venv
+ENV PATH="/opt/venv/bin:$PATH"
+
+ENV SHELL /bin/bash
+
+# todo: handle uids
+RUN useradd \
+    --create-home \
+    --no-log-init \
+    --shell /bin/bash \
+    --gid 0 \
+    --uid {uid} \
+    {user}
+
+WORKDIR {workdir}
+RUN chown {user} {workdir}
+
+# add env vars
+{env_vars}
+
+# make artifacts cache dir unrelated to build
+RUN mkdir -p {workdir}/.cache && chown -R {uid} {workdir}/.cache
+
+# copy code/etc
+COPY --chown={user} src/ {workdir}
+
+# todo handle local installs
+
+USER {user}
+
+ENV PYTHONUNBUFFERED=1
+
+ENTRYPOINT {command_arr}
+"""
+
+PYTHON_SETUP_TEMPLATE = """
+FROM {py_base_image} as base
+"""
+
+CUDA_SETUP_TEMPLATE = """
+FROM {cuda_base_image} as base
+RUN apt-get update -qq && apt-get install -y software-properties-common && add-apt-repository -y ppa:deadsnakes/ppa
+
+# install python
+# todo support runtime.txt, setup.py
+RUN apt-get update -qq && apt-get install --no-install-recommends -y \
+    {python_packages} \
+    && apt-get -qq purge && apt-get -qq clean \
+    && rm -rf /var/lib/apt/lists/*
+
+# make sure `python` points at the right version
+RUN update-alternatives --install /usr/bin/python python /usr/bin/python{py_version} 1 \
+    && update-alternatives --install /usr/local/bin/python python /usr/bin/python{py_version} 1
+"""
+
+
+def get_current_python_version():
+    full_version = sys.version.split()[0].split(".")
+    major = full_version[0]
+    version = ".".join(full_version[:2]) if len(full_version) >= 2 else major + ".0"
```

Comment:
Is this just dropping the + for the dev versions of python potentially?

Reply:
dropping the + would be correct since this is getting the version mostly to get the right docker base image, and i don't think they have them for dev versions